components/autocalc/autocalc.cpp

```diff
@@ -0,0 +1,209 @@
+            const ESM::Race *race, const ESM::Class *class_, int level, StatsBase& stats, StoreCommon *store)
+    {
+        // race bonus
+        bool male = (npc->mFlags & ESM::NPC::Female) == 0;
+
+        for (int i=0; i<ESM::Attribute::Length; ++i)
+        {
+            const ESM::Race::MaleFemale& attribute = race->mData.mAttributeValues[i];
+            stats.setAttribute(i, male ? attribute.mMale : attribute.mFemale);
+        }
+
+        // class bonus
+        for (int i=0; i<2; ++i)
+        {
+            int attribute = class_->mData.mAttribute[i];
+            if (attribute>=0 && attribute<ESM::Attribute::Length)
+                stats.setAttribute(attribute, stats.getBaseAttribute(attribute) + 10);
+            // else log an error?
+        }
+
+        // skill bonus
+        for (int attribute=0; attribute < ESM::Attribute::Length; ++attribute)
+        {
+            float modifierSum = 0;
+
+            for (int j=0; j<ESM::Skill::Length; ++j)
+            {
+                const ESM::Skill* skill = store->findSkill(j);
+
+                if (skill->mData.mAttribute != attribute)
+                    continue;
+
+                // is this a minor or major skill?
+                float add=0.2f;
+                for (int k=0; k<5; ++k)
+                {
+                    if (class_->mData.mSkills[k][0] == j)
+                        add=0.5;
+                }
+                for (int k=0; k<5; ++k)
+                {
+                    if (class_->mData.mSkills[k][1] == j)
+                        add=1.0;
+                }
+                modifierSum += add;
+            }
+            stats.setAttribute(attribute,
+                    std::min(round_ieee_754(stats.getBaseAttribute(attribute) + (level-1) * modifierSum), 100) );
+        }
+    }
+
+    /**
+     * @brief autoCalculateSkills
+     *
+     * Skills are calculated with following formulae ( http://www.uesp.net/wiki/Morrowind:NPCs#Skills ):
+     *
+     * Skills: (Level - 1) × (Majority Multiplier + Specialization Multiplier)
+     *
+     *         The Majority Multiplier is 1.0 for a Major or Minor Skill, or 0.1 for a Miscellaneous Skill.
+     *
+     *         The Specialization Multiplier is 0.5 for a Skill in the same Specialization as the class,
+     *         zero for other Skills.
+     *
+     * and by adding class, race, specialization bonus.
+     */
+    void autoCalcSkillsImpl (const ESM::NPC* npc,
+            const ESM::Race *race, const ESM::Class *class_, int level, StatsBase& stats, StoreCommon *store)
+    {
+        for (int i = 0; i < 2; ++i)
+        {
+            int bonus = (i==0) ? 10 : 25;
+
+            for (int i2 = 0; i2 < 5; ++i2)
+            {
+                int index = class_->mData.mSkills[i2][i];
+                if (index >= 0 && index < ESM::Skill::Length)
+                {
+                    stats.setBaseSkill (index, stats.getBaseSkill(index) + bonus);
```

Comment:
This relies on all skills being initialized to ```0``` before the autocalc. I know that's the case in OpenMW, but not sure about OpenCS.

Better would be to have the autocalc overwrite the skills, rather than adding. Then we can also get rid of the getBaseSkill function.

For reference: https://wiki.openmw.org/index.php?title=Research:Stats_and_Levelling#Skills